### PR TITLE
Cotel day5

### PR DESCRIPTION
**Pull request by Cotel**
# Cotel's solution for day 5 challenge

## Description

Today's problem is also easy. We need to evaluate a list of "instructions" and count the steps until one of them sends the pointer out of the bouds

## Solution

We can solve today's problem recursively too, although part 2 takes around 6 minutes in my MacBook Pro to get solved 😱 

```scala
@tailrec def instructionsUntilOut(instructions: List[Int], i: Int = 0, steps: Int = 0): Int = {
    instructions.lift(i) match {
      case None => steps
      case Some(x) => instructionsUntilOut(instructions.updated(i, x + 1), i + x, steps + 1)
      case Some(0) => instructionsUntilOut(instructions.updated(i, 1), i + 1, steps + 1)
    }
  }
```

Soo, let me explain this:

1. We pass through parameters the instructions list, the current pointer and the steps accumulator
2. We get `instructions[i]` but we lift it into `Option[Int]` because it may be out of bounds
3. We perform pattern matching for going deeper or exiting the recursion.

Note: `instructions.updated(i, n)` returns a new list with the `i` element replaced by `n`.

The solution for part 2 is the same, but it changes some operations like `i+x`. I am going to refactor a little bit for improving reutilization.

## Tests battery

Soo I write some more tests than provided because the problem wasn't clear to me at first. Writing these tests helped me to understand the problem completely.

```scala
test("For instructions [0, 3, 0, 1, -3] pointer will be out in 5 steps") {
  instructionsUntilOut(List(0, 3, 0, 1, -3)) shouldBe 5
}

test("For instructions [0, 0] pointer will be out in 4 steps") {
  instructionsUntilOut(List(0, 0)) shouldBe 4
}

test("For instructions [1, 0, -5] pointer will be out in 4 steps") {
  instructionsUntilOut(List(1, 0, -5)) shouldBe 4
}

test("For instructions [1, 1] pointer will be out in 2 steps") {
  instructionsUntilOut(List(1,1)) shouldBe 2
}
```

Happy Advent of Code!

### CODE REVIEW

File: cotel/src/main/scala/day5/Trampolines.scala [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package day5
+
+import scala.annotation.tailrec
+import scala.io.Source
+
+object Trampolines {
+
+  @tailrec def instructionsUntilOut(instructions: List[Int], i: Int = 0, steps: Int = 0)(f: Int => Int): Int = {
```
Comment (tonilopezmr) [inline cotel/src/main/scala/day5/Trampolines.scala:8]: that's the key, defaults values for not create a secondary method which initialize the variables with 0

File: cotel/src/main/scala/day5/Trampolines.scala [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      case Some(0) => instructionsUntilOut(instructions.updated(i, 1), i + 1, steps + 1)(f)
+    }
+  }
+
+  def solveWithInput(filename: String = "Day5Input.txt"): Int = {
+    val path = getClass.getClassLoader.getResource(filename).getPath
+    val instructions = (for (line <- Source.fromFile(path).getLines())
+      yield line.toInt).toList
+
+    instructionsUntilOut(instructions) { _+1 }
```
Comment (tonilopezmr) [inline cotel/src/main/scala/day5/Trampolines.scala:21]: haha 👍

**Review by tonilopezmr — APPROVED** [review]
Very good!

File: cotel/src/main/scala/day5/Trampolines.scala (hunk 1 above)
Comment (Cotel) [reply cotel/src/main/scala/day5/Trampolines.scala:8]: Preferably this is the way to go, unless the input needs some preparation in which case I make a internal function for the recursion and then I transform the input and call the recursion.

You could directly accept the tranformed input as input, but these methods has no parent above them who can prepare the calls so 🤷‍♂️

**Pull request merged**